open_spiel/algorithms/alpha_zero_torch/model.cc

```diff
@@ -371,15 +384,10 @@ std::vector<torch::Tensor> ModelImpl::forward_(torch::Tensor x,
       }
     }
   } else if (this->nn_model_ == "mlp") {
-    for (int i = 0; i < num_torso_blocks_ + 2; i++) {
-      if (i == 0) {
-        x = layers_[i]->as<torch::nn::Linear>()->forward(x);
-      } else if (i >= num_torso_blocks_ + 1) {
-        output = layers_[i]->as<MLPOutputBlockImpl>()->forward(x, mask);
-      } else {
-        x = layers_[i]->as<torch::nn::Linear>()->forward(x);
-      }
+    for (int i = 0; i < num_torso_blocks_; i++) {
+        x = layers_[i]->as<MLPTorsoBlock>()->forward(x);
     }
+    output = layers_[num_torso_blocks_]->as<MLPOutputBlockImpl>()->forward(x, mask);
```

Comment:
Note that if the above comment is accepted, then we would need to change the implementation of this `forward_` method:

```suggestion
    x = layers_[0]->as<MLPTorsoBlock>()->forward(x);
    for (int i = 1; i < num_torso_blocks_ + 1; i++) {
        x = layers_[i]->as<MLPTorsoBlock>()->forward(x);
    }
    output = layers_[num_torso_blocks_]->as<MLPOutputBlockImpl>()->forward(x, mask);
```

Reply:
Thanks for the comments.  Oops - I did it this way because I thought I was matching the python version, but looking again (https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/algorithms/alpha_zero/model.py#L225) I see that the python version is cascading nn_depth layers after an input layer, so... yes, your suggestion does seem more consistent.  Will try to commit a patch soon

Reply:
Oh wait, I guess I'm confused:  In the python code, 'observations' is a tf,placeholder, which I think means it's not a layer, just a specification of the input size. Similar for the python 'conv2d' net (not yet implemented for C++ torch)  And then 'resnet' does have a separate input block.   Anyway, I'm indifferent about this - will still plan to commit a patch (for C++), as I do think your suggestion seems natural.  But then the python might need a separate update to be consistent?

Reply:
So sorry for the late reply! I had little reception this past weekend. Thanks for your comment, I certainly see your point about consistency with the Python version.

However, for consistent meaning of the hyper-parameter across LibTorch models, and because of how the "legacy" non-residual model was implemented in #776 (i.e with the number of torso layers not including the input and output layers), let's go with the suggestions mentioned here (I see you have already implemented them in faca14c, thanks so much!).

Reply:
Thanks!  The requested changes have been submitted, so marking this resolved